cloud/services/firewalls.go

```diff
@@ -0,0 +1,316 @@
+
+		return nil, errDuplicateFirewalls
+	}
+
+	// build out the firewall rules for create or update
+	fwConfig, err := processACL(firewall, []string{clusterUID})
+	if err != nil {
+		logger.Info("Failed to process ACL", "error", err.Error())
+
+		return nil, err
+	}
+
+	if len(linodeFWs) == 0 {
+		logger.Info(fmt.Sprintf("Creating firewall %s", firewall.Name))
+		linodeFW, err = createFirewall(ctx, linodeClient, fwConfig)
+		if err != nil {
+			logger.Info("Failed to create firewall", "error", err.Error())
+
+			return nil, err
+		}
+	} else {
+		logger.Info(fmt.Sprintf("Updating firewall %s", firewall.Name))
+
+		linodeFW = &linodeFWs[0]
+		if err = updateFirewall(ctx, linodeClient, linodeFW, clusterUID, fwConfig); err != nil {
+			logger.Info("Failed to udpate firewall", "error", err.Error())
+
+			return nil, err
+		}
+	}
+
+	// Need to make sure the firewall is appropriately enabled or disabled after
+	// create or update and the tags are properly set
+	var status linodego.FirewallStatus
+	if firewall.Spec.Enabled {
+		status = linodego.FirewallEnabled
+	} else {
+		status = linodego.FirewallDisabled
+	}
+	if _, err = linodeClient.UpdateFirewall(
+		ctx,
+		linodeFW.ID,
+		linodego.FirewallUpdateOptions{
+			Status: status,
+			Tags:   util.Pointer([]string{clusterUID}),
+		},
+	); err != nil {
+		logger.Info("Failed to update Linode Firewall status and tags", "error", err.Error())
+
+		return nil, err
+	}
+
+	return linodeFW, nil
+}
+
+func createFirewall(
+	ctx context.Context,
+	linodeClient *linodego.Client,
+	fwConfig *linodego.FirewallCreateOptions,
+) (linodeFW *linodego.Firewall, err error) {
+	if linodeFW, err = linodeClient.CreateFirewall(ctx, *fwConfig); err != nil {
+		// Already exists is not an error
+		apiErr := linodego.Error{}
+		if errors.As(err, &apiErr) && apiErr.Code != http.StatusFound {
+			return nil, err
+		}
+	}
+
+	return linodeFW, nil
+}
+
+func updateFirewall(
+	ctx context.Context,
+	linodeClient *linodego.Client,
+	linodeFW *linodego.Firewall,
+	clusterUID string,
+	fwConfig *linodego.FirewallCreateOptions,
+) error {
+	if !slices.Contains(linodeFW.Tags, clusterUID) {
+		err := fmt.Errorf(
+			"firewall %s is not associated with cluster UID %s. Owner cluster is %s",
+			linodeFW.Label,
+			clusterUID,
+			linodeFW.Tags[0],
+		)
+
+		return err
+	}
+
+	if _, err := linodeClient.UpdateFirewallRules(ctx, linodeFW.ID, fwConfig.Rules); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fetch Firewalls returns all Linode firewalls with a label matching the CAPL Firewall name
+func fetchFirewalls(ctx context.Context, name string, linodeClient linodego.Client) (firewalls []linodego.Firewall, err error) {
+	var linodeFWs []linodego.Firewall
+	filter := map[string]string{"label": name}
+
+	rawFilter, err := json.Marshal(filter)
+	if err != nil {
+		return nil, err
+	}
+	if linodeFWs, err = linodeClient.ListFirewalls(ctx, linodego.NewListOptions(1, string(rawFilter))); err != nil {
```

Comment:
can we use `CreateLinodeAPIFilter` to generate the filter instead of building a raw filter?